neqo-transport/src/connection.rs

```diff
@@ -3223,4 +3249,214 @@ mod tests {
+
+        // Try to send a lot of data
+        assert_eq!(client.stream_create(StreamType::UniDi).unwrap(), 2);
+        assert_eq!(client.stream_send(2, &[0xaa; 100_000]).unwrap(), 65535);
+
+        qinfo!("client");
+        let mut c_tx_dgrams = Vec::new();
+        loop {
+            let pkt = client.process(None, now);
+            if pkt.as_dgram_ref().is_some() {
+                c_tx_dgrams.push(pkt.dgram().unwrap())
+            } else {
+                assert!(matches!(pkt, Output::Callback(_)));
+                break;
+            }
+        }
+
+        assert_eq!(c_tx_dgrams.len(), 11);
+    }
+
+    #[test]
+    #[allow(clippy::cognitive_complexity)]
+    fn congestion_control() {
+        let mut client = default_client();
+        let mut server = default_server();
+        let mut srv_buf = [0; 1_000_000];
+        connect(&mut client, &mut server);
+
+        let now = now();
+
+        // Try to send a lot of data
+        assert_eq!(client.stream_create(StreamType::UniDi).unwrap(), 2);
+        assert_eq!(client.stream_send(2, &[0xaa; 100_000]).unwrap(), 16383);
+
+        qinfo!("client");
+        let mut c_tx_dgrams = Vec::new();
+        loop {
+            let pkt = client.process(None, now);
+            if pkt.as_dgram_ref().is_some() {
+                c_tx_dgrams.push(pkt.dgram().unwrap())
+            } else {
+                assert!(matches!(pkt, Output::Callback(_)));
+                break;
+            }
+        }
+
+        assert_eq!(c_tx_dgrams.len(), 11);
+
+        // Recv all pkts and ACK
+        qinfo!("server");
+        for c_dgram in c_tx_dgrams {
+            server.process_input(c_dgram, now)
+        }
+        assert_eq!(server.stream_recv(2, &mut srv_buf).unwrap(), (12283, false));
+        let mut s_tx_dgrams = Vec::new();
+        while let Output::Datagram(dg) = server.process_output(now) {
+            s_tx_dgrams.push(dg);
+        }
+
+        // ACK pkts 0-10
+        assert_eq!(s_tx_dgrams.len(), 1);
+
+        // Handle ack, send more
+        qinfo!("client");
+        for s_dgram in s_tx_dgrams {
+            client.process_input(s_dgram, now)
+        }
+
+        assert_eq!(client.stream_send(2, &[0xab; 100_000]).unwrap(), 49152);
+
+        let mut c_tx_dgrams = Vec::new();
+        while let Output::Datagram(dg) = client.process_output(now) {
+            c_tx_dgrams.push(dg);
+        }
+
+        assert_eq!(c_tx_dgrams.len(), 21);
+
+        // Recv all pkts and ACK
+        qinfo!("server");
+        for c_dgram in c_tx_dgrams {
+            server.process_input(c_dgram, now)
+        }
+        assert_eq!(server.stream_recv(2, &mut srv_buf).unwrap(), (24617, false));
+        let mut s_tx_dgrams = Vec::new();
+        while let Output::Datagram(dg) = server.process_output(now) {
+            s_tx_dgrams.push(dg);
+        }
+
+        // ACK pkts 0-31
+        assert_eq!(s_tx_dgrams.len(), 1);
+
+        // Handle ack, send more
+        qinfo!("client");
```

Comment:
These could probably be more informative.  How about you put the content of the comment in the log?

Reply:
This is gone.